File: lsif-java/src/main/scala/com/sourcegraph/io/AbsolutePath.scala
package com.sourcegraph.io

import java.nio.file.Path
import java.nio.file.Paths

object AbsolutePath {
  def systemWorkingDirectory: Path = Paths.get(".").toAbsolutePath.normalize()
  def of(path: Path): Path = of(path, systemWorkingDirectory)
  def of(path: Path, cwd: Path): Path =
    if (path.isAbsolute)
      path
    else if (cwd.isAbsolute)
      cwd.resolve(path)
    else
      systemWorkingDirectory.resolve(cwd).resolve(path)
}

File: scip-java/src/main/kotlin/com/sourcegraph/io/AbsolutePath.kt
package com.sourcegraph.io

import java.nio.file.Path
import java.nio.file.Paths

object AbsolutePath {
    fun systemWorkingDirectory(): Path = Paths.get(".").toAbsolutePath().normalize()

    fun of(path: Path): Path = of(path, systemWorkingDirectory())

    fun of(path: Path, cwd: Path): Path =
        when {
            path.isAbsolute -> path
            cwd.isAbsolute -> cwd.resolve(path)
            else -> systemWorkingDirectory().resolve(cwd).resolve(path)
        }
}
